### PR TITLE
drivers: regulator: npm1300: Add boot state detection

### PR DESCRIPTION
**Pull request by aasinclair**
Added boot state detection, so that the reference count is set correctly and enable/disable work as expected.

With this fix it is no longer necessary to set
regulator-boot-on when the regulator is enabled in hardware.

### CODE REVIEW

**Review by gmarull — DISMISSED** [review]

File: drivers/regulator/regulator_npm1300.c [hunk 1]
```diff
@@ -512,16 +519,55 @@ int regulator_npm1300_common_init(const struct device *dev)
...
+	const struct regulator_npm1300_config *config = dev->config;
+	uint8_t data;
+
+	int ret = mfd_npm1300_reg_read(config->mfd, base, offset, &data);
+
+	if (ret == 0) {
+		*enabled = (data & mask) != 0U;
+	}
+
+	return ret;
```
Comment (fabiobaltieri) [inline drivers/regulator/regulator_npm1300.c:534]: Could you swap the condition around and have the success case inline instead and have an explicit `return 0;` at the end? Makes it dead obvious to understand imo, that would be:

```c
int ret = mfd_npm1300_reg_read(config->mfd, base, offset, &data);
if (ret != 0) {
        return ret;
}

*enabled = (data & mask) != 0U; 

return 0;
```

File: drivers/regulator/regulator_npm1300.c [hunk 2]
```diff
@@ -512,16 +519,55 @@ int regulator_npm1300_common_init(const struct device *dev)
...
+		return get_enabled_reg(dev, LDSW_BASE, LDSW_OFFSET_STATUS, LDSW2_ON_MASK, enabled);
+	default:
+		return 0;
+	}
+}
+
 int regulator_npm1300_init(const struct device *dev)
 {
 	const struct regulator_npm1300_config *config = dev->config;
+	bool enabled = false;
```
Comment (fabiobaltieri) [inline drivers/regulator/regulator_npm1300.c:558]: no need to initialize this one

**Review by fabiobaltieri — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

File: drivers/regulator/regulator_npm1300.c [hunk 3]
```diff
@@ -512,16 +519,57 @@ int regulator_npm1300_common_init(const struct device *dev)
...
+	case NPM1300_SOURCE_BUCK1:
+		return get_enabled_reg(dev, BUCK_BASE, BUCK_OFFSET_STATUS, BUCK1_ON_MASK, enabled);
+	case NPM1300_SOURCE_BUCK2:
+		return get_enabled_reg(dev, BUCK_BASE, BUCK_OFFSET_STATUS, BUCK2_ON_MASK, enabled);
+	case NPM1300_SOURCE_LDO1:
+		return get_enabled_reg(dev, LDSW_BASE, LDSW_OFFSET_STATUS, LDSW1_ON_MASK, enabled);
+	case NPM1300_SOURCE_LDO2:
+		return get_enabled_reg(dev, LDSW_BASE, LDSW_OFFSET_STATUS, LDSW2_ON_MASK, enabled);
+	default:
+		return -ENODEV;
```
Comment (gmarull) [inline drivers/regulator/regulator_npm1300.c:553]: note: this is unreachable code in practice

Comment (fabiobaltieri) [reply drivers/regulator/regulator_npm1300.c:553]: I guess it's needed to keep the compiler happy.

Comment (fabiobaltieri) [reply drivers/regulator/regulator_npm1300.c:553]: yeah the way is used looks like an enum, ah well all good, let's go!

**Pull request merged**